### PR TITLE
[#7] Replace Container with Foldable

### LINKED ISSUE #7: Remove `Container` type class and reexport `Foldable` properly
Regarding `Foldable`:

1. Custom error for `elem`
2. Don't reexport `foldl1`, `foldr1`, `maximum`, `minimum`
3. Implement `maximum` and `minimum` properly (strict and safe)

### PR DESCRIPTION
**Pull request by ChShersh**
Resolves #7 

Please, review carefully... Maybe I forgot copyright notice somewhere or some documentation.

### CODE REVIEW

**Comment by ChShersh** [comment]
@vrom911 This is now ready to be reviewed! Please, review and share your thoughts 🙂

File: README.md [hunk 1]
```diff
@@ -136,18 +135,8 @@ Gotchas [↑](#structure-of-this-tutorial)
...
-  [`Container` type class](https://github.com/serokell/universum/blob/b6353285859e9ed3544bddbf55d70237330ad64a/src/Universum/Container/Class.hs#L180)
-  but that restricts the usage of functions like `length` over `Maybe`, `Either`, `Identity` and tuples.
-  We're also using _GHC 8_ feature of
-  [custom compile-time errors](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#custom-compile-time-errors)
-  to produce
-  [more helpful messages](https://github.com/serokell/universum/blob/54a742c10720f11c739f2d268365d723924b83a9/src/Containers.hs#L474).
-* As a consequence of previous point, some functions like `traverse_`, `forM_`, `sequenceA_`, etc.
-  are generalized over `Container` type classes.
+* You can't call `elem` and `notElem` functions over `Set` and `HashSet`. These
+  functions were forbidden for these two types because of performance reasons.
```
Comment (vrom911) [inline README.md]: Probably *are* instead of *were*?

File: src/Universum/Base.hs [hunk 2]
```diff
@@ -15,14 +22,13 @@ module Universum.Base
...
        , module Data.Eq
-       , module Data.Foldable
        , module Data.Ord
-       , module Data.Traversable
 
          -- * System IO
        , module System.IO
 
          -- * Base GHC types
+       , module Data.Coerce
```
Comment (vrom911) [inline src/Universum/Base.hs]: That's very useful!

File: src/Universum/Foldable.hs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+-- | This module exports all 'Foldable'-related and 'Traversable'-related stuff.
```
Comment (vrom911) [inline src/Universum/Foldable.hs]: 'Foldable' and 'Traversable' related stuff

File: src/Universum/Foldable/Fold.hs [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+{-# SPECIALIZE orM  :: [IO Bool] -> IO Bool #-}
+{-# SPECIALIZE anyM :: (a -> IO Bool) -> [a] -> IO Bool #-}
+{-# SPECIALIZE allM :: (a -> IO Bool) -> [a] -> IO Bool #-}
+
+----------------------------------------------------------------------------
+-- Type level tricks
+----------------------------------------------------------------------------
+
+#if __GLASGOW_HASKELL__ >= 800
+type family DisallowElem (f :: Type -> Type) :: Constraint where
```
Comment (vrom911) [inline src/Universum/Foldable/Fold.hs]: Cool! 😎

File: README.md [hunk 5]
```diff
@@ -136,18 +135,8 @@ Gotchas [↑](#structure-of-this-tutorial)
...
 * `error` takes `Text`.
```
Comment (vrom911) [inline README.md]: Can you please also check `Container` usage in migration section of this document?

**Pull request merged**